Template/Trigger/ElementVisibilityTrigger.web.js

```diff
@@ -310,10 +294,86 @@
             return false;
         }
 
+        function setMutationObserver(triggerEvent) {
+            return function () {
+                if (observeDomChanges && isMutationObserverSupported) {
+                    var targetNode = documentAlias.body;
+                    var config = {attributes: true, childList: true, subtree: true};
+                    observerMutation = new MutationObserver(function (mutationsList) {
+                        mutationObserverCallback(mutationsList, triggerEvent);
+                    });
+                    observerMutation.observe(targetNode, config);
+                }
+            };
+        }
+
+        function mutationObserverCallback(mutationsList, triggerEvent) {
+            for (var index in mutationsList) {
+                var mutation = mutationsList[index];
+                if (mutation.type === 'childList' || mutation.type === 'attributes') {
+                    var addedNodes = mutation.addedNodes;
+                    addedNodes.forEach(function (node) {
+                        TagManager.dom.bySelector(selectors).forEach(function (element) {
+                            if (node.contains(element)) {
+                                if (blockTrigger || (onlyOncePerElement && isNodeEventTriggered(element))) {
+                                    return;
+                                }
+
+                                if (!isNodeInViewport(element) && observerIntersection) {
+                                    observerIntersection.observe(element);
+
+                                    return;
+                                }
+
+                                var percentVisible = Math.max(getPercentVisible(element), minPercentVisible);
+                                commonTrigger(triggerEvent, percentVisible, element);
+                                if (fireTriggerWhen === 'oncePage') {
```

Comment:
fyi we have the similar code here in multiple places. Also around `if (self.scrollIndex) { TagManager.window.offScroll(self.scrollIndex);` etc. Maybe could be reused by adding a method?

Reply:
Yes, I have moved to a common function and should cover all cases, please do look for any case I would have missed